xrspatial/experimental/polygonize.py

```diff
@@ -0,0 +1,590 @@
+    from shapely.geometry import Polygon
+
+    # Convert list of point arrays to shapely Polygons.
+    polygons = list(map(
+        lambda points: Polygon(points[0], points[1:]), polygon_points))
+
+    df = gpd.GeoDataFrame({column_name: column, "geometry": polygons})
+    return df
+
+
+def _to_spatialpandas(
+    column: List[Union[int, float]],
+    polygon_points: List[np.ndarray],
+    column_name: str,
+):
+    from spatialpandas import GeoDataFrame
+    from spatialpandas.geometry import PolygonArray
+
+    # spatialpandas expects 1d numpy arrays.
+    for i, arrays in enumerate(polygon_points):
+        polygon_points[i] = \
+            list(map(lambda array: np.reshape(array, -1), arrays))
+
+    df = GeoDataFrame({
+        column_name: column, "geometry": PolygonArray(polygon_points)})
+    return df
+
+
+def _polygonize_numpy(
+    values: np.ndarray,
+    mask: Optional[np.ndarray],
+    connectivity_8: bool,
+    transform: Optional[np.ndarray],
+    column_name: str,
+    return_type: str,
+):
+    ny, nx = values.shape
+    if nx == 1:
+        # Algorithm requires nx > 1 to differentiate between facing E
+        # (forward == 1) and facing N (forward == nx), so add extra column to
+        # values array and mask the column out.
+        nx = 2
+        values = np.hstack((values, np.empty_like(values)))
+        if mask is not None:
+            mask = np.hstack((mask, np.zeros_like(mask)))
+        else:
+            mask = np.zeros_like(values, dtype=bool)
+            mask[:, 0] = True
+
+    values = values.ravel()
+    if mask is not None:
+        mask = mask.ravel()
+
+    column, polygon_points = _scan(
+        values, mask, connectivity_8, transform, nx, ny)
+
+    if return_type == "numpy":
+        return column, polygon_points
+    elif return_type == "awkward":
+        return _to_awkward(column, polygon_points)
+    elif return_type == "geopandas":
+        return _to_geopandas(column, polygon_points, column_name)
+    elif return_type == "spatialpandas":
+        return _to_spatialpandas(column, polygon_points, column_name)
+    else:
+        raise ValueError(f"Invalid return_type '{return_type}'")
+
+
+def polygonize(
+    raster: xr.DataArray,                  # shape (ny, nx) integer or float
+    mask: Optional[xr.DataArray] = None,   # shape (ny, nx) bool/integer/float
+    connectivity: int = 4,                 # 4 or 8
+    transform: Optional[np.ndarray] = None,  # shape (6,)
+    column_name: str = "DN",
+    return_type: str = "numpy",
+):
+    """
+    Polygonize creates vector polygons for connected regions of pixels in a
+    raster that share the same pixel value.  It is a raster to vector
+    converter.
+
+    Parameters
+    ----------
+    raster: xr.DataArray
+        Input raster.
+
+    mask: xr.DataArray, optional
+        Optional input mask.  Pixels to include should have mask values of 1
+        or True, pixels to exclude should have 0 or True.  This is the
```

Comment:
I think it should be "pixels to exclude should have 0 or False".

Reply:
Good catch!